modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/TransactionProxyImpl.java

```diff
@@ -421,4 +421,63 @@ private void saveFuture(IgniteInternalFuture<IgniteInternalTx> fut) {
     @Override public String toString() {
         return S.toString(TransactionProxyImpl.class, this);
     }
+
+    /** {@inheritDoc} */
+    @Override public void savepoint(String name) throws IllegalArgumentException {
+        savepoint(name, false);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void savepoint(String name, boolean overwrite) throws IllegalArgumentException {
+        if (implicit())
+            throw new IllegalStateException("Savepoints can be used only inside explicit transactions.");
+
+        enter();
+
+        try {
+            tx.savepoint(name, overwrite);
+        }
+        catch (IgniteCheckedException e) {
+            throw U.convertException(e);
+        }
+        finally {
+            leave();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void rollbackToSavepoint(String name) {
```

Comment:
probably we should add rollbackToSavepointAsync method

Reply:
No, we shouldn't start any tx/cache action simultaneously with saving current state. For example, if we simultaneously put something to new key - we can't be sure that this key will be saved in savepoint nor it not saved.